### DISCUSSION TITLE
Difference with other "delegate" crates.

### DISCUSSION THREAD

**Discussion by bragov4ik** [SOF3/portrait#13] [Q&A] [answered]
I stumbled upon this crate and I think it would be nice to have this written down somewhere. I think the list should simplify choosing between the crates & better understand the aim of this crate.

In particular, in discussions about rust delegate rfc's I found [delegate](https://crates.io/crates/delegate) and [ambassador](https://crates.io/crates/ambassador).

So, the question is: what are the features that are unique to this crate?

**Comment by bragov4ik** [comment, marked as answer]
I discovered one use case this crate excels at. Apparently, other crates cannot elegantly delegate associated methods to some other type. For example, the following trait delegate implementations benefit from `portrait`:
```rust
trait SomeTrait {
    fn foo() -> i32;
    fn baz();
}
```
- [ambassador does not support methods without receiver](https://github.com/hobofan/ambassador/issues/15)
- [delegate does not instrument the trait](https://github.com/Kobzol/rust-delegate/issues/27#issuecomment-742602478), so implementations have to list all trait methods (which kinda defeats the purpose of macro: getting rid of code duplication)

So, portrait can solve such use case easily:
```rust
struct BaseCase;

impl SomeTrait for BaseCase {
    fn foo() -> i32 { 123 }
    fn baz() {
        println!("Hello, World!");
    }
}

struct Wrapper<T: SomeTrait>(PhantomData<T>);

#[portrait::fill(portrait::delegate(T))]
impl<T: SomeTrait> SomeTrait for Wrapper<T> { }
```

Thanks for making this crate; now my inner perfectionist is happy 🎉

**Comment by SOF3** [comment]
portrait is actually a framework for exposing trait body to proc macros, which includes other fillers like default and log. I added delegate simply because it happens to fit in the framework as well; didn't really do research on other delegation crates in the ecosystem. But much thanks for the comparison!
